toolkit/tools/pkg/licensecheck/licensecheck.go

```diff
@@ -0,0 +1,411 @@
+	}
+	if pedantic && len(result.DuplicatedDocs) > 0 {
+		sb.WriteString(fmt.Sprintf("\tduplicated license files:\n\t\t%s\n", strings.Join(result.DuplicatedDocs, "\n\t\t")))
+	}
+	return sb.String()
+}
+
+// runLicenseCheckInChroot searches for bad licenses amongst the RPMs mounted into the chroot. This function is meant
+// to be called from inside the chroot's context.
+func (l *LicenseChecker) runLicenseCheckInChroot() (results []LicenseCheckResult, err error) {
+	const searchReportIntervalPercent = 10
+
+	// Find all the rpms in the chroot
+	rpmsToSearchPaths, err := l.findRpmPaths()
+	if err != nil {
+		return nil, fmt.Errorf("failed to walk srpm directory:\n%w", err)
+	}
+	if len(rpmsToSearchPaths) == 0 {
+		logger.Log.Warnf("No rpms found in %s", l.simpleToolChroot.ChrootRelativeMountDir())
+		return nil, nil
+	}
+
+	// Scan each rpm in parallel
+	resultsChannel := make(chan LicenseCheckResult, len(rpmsToSearchPaths))
+	cancel := make(chan struct{})
+	logger.Log.Infof("Queuing %d rpms for search", len(rpmsToSearchPaths))
+	l.queueWorkers(rpmsToSearchPaths, resultsChannel, cancel)
+	logger.Log.Infof("Searching rpms...")
+
+	// Wait for all the workers to finish, updating the progress as results come in
+	numProcessed := 0
+	lastReportPercent := 0
+	for range rpmsToSearchPaths {
+		result := <-resultsChannel
+		if result.err != nil {
+			// Signal the workers to stop if there is an error
+			err = fmt.Errorf("failed to search srpm:\n%w", result.err)
+			close(cancel)
+			return
+		}
+		numProcessed++
+		percentProcessed := (numProcessed * 100) / len(rpmsToSearchPaths)
+		if percentProcessed-lastReportPercent >= searchReportIntervalPercent {
+			logger.Log.Infof("Checked %d/%d rpms (%d%%)", numProcessed, len(rpmsToSearchPaths), percentProcessed)
+			lastReportPercent = percentProcessed
+		}
+		results = append(results, result)
+	}
+	return
+}
+
+// findRpmPaths walks the chroots's mount directory to find all *.rpm files. The paths are returned relative to the
+// chroot's root.
+func (s *LicenseChecker) findRpmPaths() (foundSrpmPaths []string, err error) {
+	const rpmExtension = ".rpm"
+	err = filepath.Walk(s.simpleToolChroot.ChrootRelativeMountDir(), func(path string, info os.FileInfo, walkErr error) error {
+		if walkErr != nil {
+			return walkErr
+		}
+		if info.IsDir() {
+			return nil
+		}
+		if !strings.HasSuffix(path, rpmExtension) {
+			return nil
+		}
+
+		foundSrpmPaths = append(foundSrpmPaths, path)
+		return nil
+	})
+	if err != nil {
+		err = fmt.Errorf("failed to walk directory:\n%w", err)
+		return nil, err
+	}
+	return foundSrpmPaths, nil
+}
+
+// queueWorkers queues up workers to search the RPMs in parallel. Each worker will wait on the jobSemaphore before starting.
+func (l *LicenseChecker) queueWorkers(rpmsToSearchPaths []string, resultsChannel chan LicenseCheckResult, cancel chan struct{}) {
+	for _, rpmPath := range rpmsToSearchPaths {
```

Comment:
Mostly curious: why spin up and tear down a new worker for each path instead of having X workers grabbing paths from a channel and waiting for them to finish? Since we're working with files, I don't think there's going to be any performance impact, thus only curious. Was the code cleaner?

Reply:
My personal take, it makes the code much easier to read. go routines are super lightweight so there isn't really an issue with spinning them all up right away.